schema/data.proto

```diff
@@ -0,0 +1,487 @@
+//     }
+//
+// 3) A pointer where the pointee was captured:
+//
+//     T x = { 3, 7 };
+//     T* p = &x;
+//
+//     {   // Captured variable
+//         name: "p",
+//         type: "T*",
+//         value: "0x00500500",
+//         members { name: "m1", value: "3", type: "int" },
+//         members { name: "m2", value: "7", type: "int" }
+//     }
+//
+// 4) A pointer where the pointee was not captured:
+//
+//     T* p = new T;
+//
+//     {   // Captured variable
+//         name: "p",
+//         type: "T*",
+//         value: "0x00400400"
+//         status { is_error: true, description { format: "unavailable" } }
+//     }
+//
+// The status should describe the reason for the missing value,
+// such as `<optimized out>`, `<inaccessible>`, `<pointers limit reached>`.
+//
+// Note that a null pointer should not have members.
+//
+// 5) An unnamed value:
+//
+//     int* p = new int(7);
+//
+//     {   // Captured variable
+//         name: "p",
+//         value: "0x00500500",
+//         type: "int*",
+//         members { value: "7", type: "int" } }
+//
+// 6) An unnamed pointer where the pointee was not captured:
+//
+//     int* p = new int(7);
+//     int** pp = &p;
+//
+//     {  // Captured variable
+//         name: "pp",
+//         value: "0x00500500",
+//         type: "int**",
+//         members {
+//             value: "0x00400400",
+//             type: "int*"
+//             status {
+//                 is_error: true,
+//                 description: { format: "unavailable" } }
+//             }
+//         }
+//     }
+//
+// To optimize computation, memory and network traffic, variables that
+// repeat in the output multiple times can be stored once in a shared
+// variable table and be referenced using the `var_table_index` field.  The
+// variables stored in the shared table are nameless and are essentially
+// a partition of the complete variable. To reconstruct the complete
+// variable, merge the referencing variable with the referenced variable.
+//
+// When using the shared variable table, the following variables:
+//
+//     T x = { 3, 7 };
+//     T* p = &x;
+//     T& r = x;
+//
+//     { name: "x", var_table_index: 3, type: "T" }  // Captured variables
+//     { name: "p", value "0x00500500", type="T*", var_table_index: 3 }
+//     { name: "r", type="T&", var_table_index: 3 }
+//
+//     {  // Shared variable table entry #3:
+//         members { name: "m1", value: "3", type: "int" },
+//         members { name: "m2", value: "7", type: "int" }
+//     }
+//
+// Note that the pointer address is stored with the referencing variable
+// and not with the referenced variable. This allows the referenced variable
+// to be shared between pointers and references.
+//
+// The type field is optional. The debugger agent may or may not support it.
+//
+message Variable {
+  // Name of the variable, if any.
+  string name = 1;
+
+  // Simple value of the variable.
+  string value = 2;
```

Comment:
Thoughts on having optional strings?  If I understand things properly, either this is set or the var_table_index is.

Reply:
I've gone through and marked more strings optional.